iroh/src/baomap/flat.rs

```diff
@@ -895,42 +895,65 @@ impl Store {
         Ok((tag, size))
     }
 
+    fn import_bytes_sync(&self, data: Bytes, format: BlobFormat) -> io::Result<TempTag> {
+        let temp_data_path = self.temp_path();
+        std::fs::write(&temp_data_path, &data)?;
+        let id = 0;
+        let file = ImportFile::TempFile(temp_data_path);
+        let progress = IgnoreProgressSender::default();
+        let (tag, _size) = self.finalize_import_sync(file, format, id, progress)?;
+        // we have the data in memory, so we can just insert it right now
+        if data.len() < self.0.options.inline_threshold as usize {
+            let mut state = self.0.state.write().unwrap();
+            state.data.insert(*tag.hash(), data);
+        }
+        Ok(tag)
+    }
+
     fn finalize_import_sync(
         &self,
         file: ImportFile,
         format: BlobFormat,
         id: u64,
         progress: impl ProgressSender<Msg = ImportProgress> + IdGenerator,
     ) -> io::Result<(TempTag, u64)> {
-        let complete_io_guard = self.0.complete_io_mutex.lock().unwrap();
         let size = file.path().metadata()?.len();
         progress.blocking_send(ImportProgress::Size { id, size })?;
         let progress2 = progress.clone();
         let (hash, outboard) = compute_outboard(file.path(), size, move |offset| {
             Ok(progress2.try_send(ImportProgress::OutboardProgress { id, offset })?)
         })?;
         progress.blocking_send(ImportProgress::OutboardDone { id, hash })?;
-        let (tag, new, outboard) = match file {
-            ImportFile::External(path) => {
-                use baomap::Store;
-                let tag = self.temp_tag(HashAndFormat(hash, format));
-                (tag, CompleteEntry::new_external(size, path), outboard)
-            }
-            ImportFile::TempFile(path) => {
+        use baomap::Store;
+        // from here on, everything related to the hash is protected by the temp tag
+        let tag = self.temp_tag(HashAndFormat(hash, format));
+        let hash = *tag.hash();
+        let temp_outboard_path = if let Some(outboard) = outboard.as_ref() {
+            let uuid = new_uuid();
+            // we write the outboard to a temp file first, since while it is being written it is not complete.
+            // it is protected from deletion by the temp tag.
+            let temp_outboard_path = self.0.options.partial_outboard_path(hash, &uuid);
+            std::fs::write(&temp_outboard_path, outboard)?;
+            Some(temp_outboard_path)
+        } else {
+            None
+        };
+        // before here we did not touch the complete files at all.
+        // all writes here are protected by the temp tag
+        let complete_io_guard = self.0.complete_io_mutex.lock().unwrap();
+        // move the data file into place, or create a reference to it
+        let new = match file {
```

Comment:
This was a bug before. You *must* pass through the format.

if somebody adds a hashseq using import_stream or import_reader, the resulting temp tag *must still* be with format HashSeq.